specs/phase1/data-availability-sampling.md

```diff
@@ -0,0 +1,150 @@
+| `SampleIndex` | `uint64` | A sample index, corresponding to chunk of extended data |
+| `BLSPoint` | `uint256` | A number `x` in the range `0 <= x < MODULUS` |
+
+
+## New containers
+
+### DASSample
+
+```python
+class DASSample(Container):
+    slot: Slot
+    shard: Shard
+    index: SampleIndex
+    proof: BLSKateProof
+    data: Vector[BLSPoint, POINTS_PER_SAMPLE]
+```
+
+### ShardBlob
+
+The blob of data, effectively a block. Network-only.
+
+```python
+class ShardBlob(Container):
+    # Slot and shard that this blob is intended for
+    slot: Slot
+    shard: Shard
+    # The actual data
+    data: List[BLSPoint, POINTS_PER_SAMPLE * MAX_SAMPLES_PER_BLOCK]
+```
+
+Note that the hash-tree-root of the `ShardBlob` does not match the `ShardHeader`, 
+since the blob deals with full data, whereas the header includes the Kate commitment instead.
+
+### SignedShardBlob
+
+Network-only.
+
+```python
+class ShardBlob(Container):
+    message: ShardBlob
+    signature: BLSSignature
+```
+
+
+## Helper functions
+
+### Data extension
+
+Implementations:
+- [Python](https://github.com/protolambda/partial_fft/blob/master/das_fft.py)
+- [Go](https://github.com/protolambda/go-kate/blob/master/das_extension.go)
+
+```python
+def das_fft_extension(data: Sequence[Point]) -> Sequence[Point]:
+    """Given some even-index values of an IFFT input, compute the odd-index inputs, such that the second output half is all zeroes."""
+    poly = inverse_fft(data)
+    return fft(poly + [0]*len(poly))[1::2]
+```
+
+### Data recovery
+
+See [Reed-Solomon erasure code recovery in n*log^2(n) time with FFTs](https://ethresear.ch/t/reed-solomon-erasure-code-recovery-in-n-log-2-n-time-with-ffts/3039) for theory.
+Implementations:
+- [Original Python](https://github.com/ethereum/research/blob/master/mimc_stark/recovery.py)
+- [New optimized approach in python](https://github.com/ethereum/research/tree/master/polynomial_reconstruction)
+- [Old approach in Go](https://github.com/protolambda/go-kate/blob/master/recovery.go)
+
+```python
+def recover_data(data: Sequence[Optional[Point]]) -> Sequence[Point]:
+    """Given an a subset of half or more of the values (missing values are None), recover the None values."""
+    ...
+```
+
+## DAS functions
+
+```python
+def extend_data(data: Sequence[Point]) -> Sequence[Point]:
+    # To force adjacent data into the same proofs, reverse-bit-order the whole list.
+    evens = [data[reverse_bit_order(i, len(data))] for i in range(len(data))]
+    # last step of reverse-bit-order: mix in the extended data.
+    # When undoing the reverse bit order: 1st half matches original data, and 2nd half matches the extension.
+    odds = das_fft_extension(data)
+    return [evens[i//2] if i % 2 == 0 else odds[i//2] for i in range(len(data)*2)]
+```
+
+```python
+def unextend_data(extended_data: Sequence[Point]) -> Sequence[Point]:
+    return [extended_data[reverse_bit_order(i, len(extended_data))] for i in range(len(extended_data)//2)]
```

Comment:
Why is this not just `return extended_data[:len(extended_data)//2]`? Isn't the goal of `extend_data` to put the extended chunks in the right half and keep the original chunks in the left half? Though `extend_data` as currently written seems to be trying to do something different....

Reply:
Currently the "extended_data" is assumed to be in bit-reverse order, such that taking the FFT of it, unmodified, results in the half of zeroes. And when sampling, I think it is nice to just take slices of the extended data, to avoid copies and such. 

We can pick if "extended data" means fully-prepared (bit-reverse order, odd indices being extension, etc.) or that padded form (the original data, then the extension). Without bit-reverse ordering the padded form, it cannot really be used effectively, so I suppose we want the other form?

Reply:
Fixed. If anyone else is interested, full description in https://github.com/ethereum/eth2.0-specs/pull/2172#issuecomment-753465299